core/src/main/java/io/grpc/internal/MessageFramer.java

```diff
@@ -302,7 +304,11 @@ private void releaseBuffer() {
 
   private void commitToSink(boolean endOfStream, boolean flush) {
     WritableBuffer buf = buffer;
-    buffer = null;
+    // If we have a buffer still and we are not trying to do the last flush on a close() call,
+    // try and reuse the remainder of the buffer.
+    if (buf != null && !isClosed()) {
```

Comment:
@buchgr 

One alternative way of doing this is to hold a reference to the buffer but not to early increment it's reference count. When we attempt to **use** the buffer the next time we can try and increment it's refcount and if we can't its been recycled already and we will need to get a new buffer. This should be thread safe and it doesn't introduce a memory leak risk of any significance